### DISCUSSION TITLE
Inadyn Cloudflare DDNS Help?

### DISCUSSION THREAD

**Discussion by KamiYamabushi** [troglobit/inadyn#400] [Q&A]
I realize this is for AsusWRT-Merlin but I'm unsure if I'm doing something wrong with Inadyn or Merlin (haven't gotten a response there).

So I've spent about 3 hours trying to get this working so far while following the instructions from https://github.com/RMerl/asuswrt-merlin.ng/wiki/DDNS-services and https://github.com/troglobit/inadyn and https://github.com/troglobit/inadyn/blob/master/examples/cloudflare-ipv4-ipv6.conf.

Unfortunately, I keep getting the dreaded "Request error! Please try again." message. Any advice?

My inadyn.conf (stored at /jffs - minus the sensitive information):

```
# In-A-Dyn v2.x configuration file format
period = 300
user-agent = Mozilla/5.0
allow-ipv6 = true

# ipv4
provider cloudflare.com:1 {
    # checkip-server = 1.1.1.1
    # checkip-path = /cdn-cgi/trace
    checkip-command = "/bin/nvram get wan_ipaddr"
    username = domain.com
    password = API-Token-Key # Create a unique custom api token with the following permissions: Zone.Zone - Read, Zone.DNS - Edit.
    hostname = sub.domain.com
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}

# ipv6
provider cloudflare.com:2 {
    # checkip-server = dns64.cloudflare-dns.com
    # checkip-path = /cdn-cgi/trace
    checkip-command = "/bin/nvram get ipv6_wan_addr"  | awk -F '/' '{print $1}'
    username = domain.com
    password = API-Token-Key # Use the same api token as above.
    hostname = sub.domain.com
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}
```

My ddns-start script (stored at /jffs/scripts):

```
#!/bin/sh
inadyn --once -f "/jffs/inadyn.conf" -e "/sbin/ddns_custom_updated 1" --continue-on-error "/sbin/ddns_custom_updated 0"
```

Log example of issue:

```
Dec  5 03:01:05 rc_service: httpds 30741:notify_rc restart_ddns_le
Dec  5 03:01:05 ddns: update CUSTOM , wan_unit 0
Dec  5 03:01:05 ddns: Clear ddns cache.
Dec  5 03:01:05 custom_script: Running /jffs/scripts/ddns-start (args: <redacted IP address>)
Dec  5 03:01:27 watchdog: start ddns.
Dec  5 03:01:27 rc_service: watchdog 2885:notify_rc start_ddns watchdog
Dec  5 03:01:27 ddns: update CUSTOM , wan_unit 0
Dec  5 03:01:27 ddns: Clear ddns cache.
Dec  5 03:01:27 custom_script: Running /jffs/scripts/ddns-start (args: <redacted IP address>)
Dec  5 03:01:57 watchdog: start ddns.
Dec  5 03:01:57 rc_service: watchdog 2885:notify_rc start_ddns watchdog
Dec  5 03:01:57 ddns: update CUSTOM , wan_unit 0
Dec  5 03:01:57 ddns: Clear ddns cache.
Dec  5 03:01:57 custom_script: Running /jffs/scripts/ddns-start (args: <redacted IP address>)
```

**Comment by troglobit** [comment]
I don't know what half of this is, so I'm assuming some Merlin tricks and will comment only on what I see as odd.

 1. `--continue-on-error` doesn't take an argument, maybe the Merlin ppl have a local patch, I dunno
 2. The log file doesn't resemble anything that comes from In-a-dyn
 3. The .conf file, nowever, looks like you've just pasted in the example from the README?
    - Do you have two Cloudflare domains, (:1 and :2)?  I'm guessing no
    - Have you entered your username and API key properly?
    - Have you changed the hostname to your hostname?

**Comment by KamiYamabushi** [comment]
Thanks for trying to help out anyhow!

1. I'm not sure about this one either, I just tried to follow an example they had provided.
2. No worries!
3. I combined a couple things. Part of it is from the README, the other part is from this https://github.com/troglobit/inadyn/blob/master/examples/cloudflare-ipv4-ipv6.conf
   * I don't have two domains :1 and :2 but I have a single subdomain that I'm trying to update an A record and AAAA record for IPv4 and IPv6.
   * Unless I misunderstood, in the examples I've seen, the username is the Zone.Name which I assumed was the domain name? The API Token key I copied directly from the page so it is accurate.
   * The hostname I assumed was the subdomain?

**Reply by troglobit** [reply]
Well, if you only have one domain, you don't need the :1 and :2 sections, you only need one (see below).

I've not made the Cloudflare plugin myself, and barely know how it works.  I can only refer you to the README, which states:

```
# Create a unique custom API token with the following permissions:
# -> Zone.Zone - Read, Zone.DNS - Edit.
provider cloudflare.com {
    username = zone.name
    password = api_token_important_read_comment
    hostname = hostname.zone.name
    ttl = 1 # optional, value of 1 is 'automatic'.
    proxied = false # optional.
}
```

**Reply by moosefuel** [reply]
How can we track down whoever made the cloudflare plugin for a little extra info?

**Reply by troglobit** [reply]
It was contributed by @SimonPilkington but I'm not sure if he's still using Inadyn?  One way to debug the problem is to run once in debug mode: `inadyn -1 -l debug --foreground --force`

**Reply by SimonPilkington** [reply]
I do still use Inadyn but a plethora of people have made contributions to the plugin since I wrote the initial version and I have always used a very simple configuration which is what is described in the README. I also do not know why KamiYamabushi had to set his API token to All Zones as I use a token for a specific zone and this works completely fine for me.

**Reply by troglobit** [reply]
I believe @moosefuel has a similar issue in https://github.com/troglobit/inadyn/discussions/425#discussioncomment-6473684

**Reply by SimonPilkington** [reply]
Added some answers there.

**Comment by KamiYamabushi** [comment]
Thanks for your assistance everyone, it turns out the issue is much simpler than expected and also confusing. I had to update my Cloudflare API Token from a specific zone to "All zones". I don't really understand why that is required as I'm only targeting a specific zone (and only have one zone for that matter), but it is what it is.

**Reply by troglobit** [reply]
Great to hear you got it working! This should perhaps be added to the documentation somewhere.
